### PR TITLE
Detailed Output

### PR DESCRIPTION
**Pull request by jimleeder123**
This is for the get and post routes of the detailed output. The detailed output page can be accessed by guests who have completed their profile AND have the correct session set up as well as the authenticated user to whom the related enquiry belongs to. There is a form for both types of users to get the notification sent to them. There is a form for guests only which links them to log in to the system - this is just a placeholder route for now to come in a future PR.

There is a middleware for the route that controls access. Tests have been set up for the two routes.

There is also a change to logic built previously where if an authenticated user followed the "not been here before" flow then the enquiry was not getting attached to them. They would have been sent the email and then not able to access the detailed view, so this has been amended and the test for it has been updated.

There are questions in the comments as normal.

### CODE REVIEW

**Comment by jimleeder123** [comment]
[Send detailed output email](https://trello.com/c/SAH6RNzB/35-send-detailed-output-email)

**Comment by jimleeder123** [comment]
[Present web detailed output](https://trello.com/c/T2ZcerZf/30-present-web-detailed-output)

File: resources/views/web/space-calculator/detailed-results.blade.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      <x-errors :errors="$errors" />
+      @csrf
+      <button type="submit" title="Get detailed results PDF">Get detailed results PDF</button>
+   </form>
+
+   @guest
```
Comment (gbuckingham89) [inline resources/views/web/space-calculator/detailed-results.blade.php:26]: I didn't know about these - nice! 👌

File: app/Http/Middleware/GuardSpaceCalculatorOutput.php [hunk 2]
```diff
@@ -18,8 +18,8 @@ class GuardSpaceCalculatorOutput
      */
     public function handle(Request $request, Closure $next): Response
     {
-        if (Auth::check()) { // todo: change this the prospect portal route when built
-            return redirect(route('web.space-calculator.index'));
+        if (Auth::check()) {
+            return redirect(route('web.portal.index'));
```
Comment (gbuckingham89) [inline app/Http/Middleware/GuardSpaceCalculatorOutput.php:22]: Well remembered! 😀

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        if (!Auth::check()) {
+            $user = $model->enquiry->user;
+
+            if (!$user) {
+                return redirect(route('web.space-calculator.index'));
+            }
+
+            if (!$user->has_completed_profile) {
+                return redirect(route('web.space-calculator.index'));
+            }
```
Comment (gbuckingham89) [inline app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:36]: For these, I guess we could redirect to the summary outputs instead?

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+        if (Auth::check()) {
+
+            /**
+             * @var User $user
+             */
+            $user = Auth::user();
+
+            if (!$user->has_completed_profile) {
+                // todo: discuss - would this link to a profile edit screen perhaps later?
```
Comment (gbuckingham89) [inline app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:55]: Yep in a future phase you can see project editing being useful - but is out-of-scope for the current project!

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     * @param Request $request
+     * @param Closure $next
+     * @return Response
+     */
+    public function handle(Request $request, Closure $next): Response
+    {
+        /**
+         * @var SpaceCalculatorInput $model
+         */
+        $model = $request->route()?->parameter('spaceCalculatorInput');
```
Comment (gbuckingham89) [inline app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:25]: `$model` isn't the nicest name - what model is it? Maybe `$input` would be better?

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+             * @var User $user
+             */
+            $user = Auth::user();
+
+            if (!$user->has_completed_profile) {
+                // todo: discuss - would this link to a profile edit screen perhaps later?
+                return redirect(route('web.portal.index'));
+            }
+
+            if ($model->enquiry->user_id != $user->id) {
```
Comment (gbuckingham89) [inline app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:59]: You could using the eloquent model comparison to make it a bit nicer to read;

```php
if( !$model->enquiry->user->is($user))
```

File: app/Http/Controllers/Web/SpaceCalculator/OutputsController.php [hunk 7]
```diff
@@ -118,11 +122,45 @@ public function postProfile(
...
+
+        if (!Auth::check()) {
+            SendAction::run(
+                $user,
+                $request->ip(),
+                route('web.portal.index', $spaceCalculatorInput)
+            );
+            $request->session()->flash('auth-sent-user', $user);
+            return redirect(route('web.auth.sent'));
+        }
```
Comment (gbuckingham89) [inline app/Http/Controllers/Web/SpaceCalculator/OutputsController.php:162]: I don't think we need to be doing this - we've got the separate form & endpoint for requesting & sending the magic link.

File: app/Http/Controllers/Web/SpaceCalculator/OutputsController.php [hunk 8]
```diff
@@ -118,11 +122,45 @@ public function postProfile(
...
+
+        return redirect(route('web.portal.index'));
```
Comment (gbuckingham89) [inline app/Http/Controllers/Web/SpaceCalculator/OutputsController.php:164]: After requesting detailed output email, we should redirect back to the detailed output page.

File: tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                areaTypes: collect(),
+            ));
+
+        $this->get(route('web.space-calculator.outputs.detailed', $inputs->uuid))
+            ->assertOk()
+            ->assertDontSeeText('Log in to view your portal')
+            ->assertViewIs('web.space-calculator.detailed-results');
+    }
+
+    // todo: discuss - is this too over the top for testing? Since auth user's don't have session check in middleware
```
Comment (gbuckingham89) [inline tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php:60]: I'd start by thinking about what the behaviour should be in that situation. 

Hopefully, it's not even possible to get into that state.

So in this case, yes, I probably wouldn't have this test.

File: tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        $this->authenticateUser($user);
+
+        $this->mock(Calculator::class)
+            ->shouldNotHaveBeenCalled();
+
+        $this->get(route('web.space-calculator.outputs.detailed', $inputs->uuid))
+            ->assertRedirect(route('web.portal.index'));
+    }
+
+    // todo: discuss - this one possibly over the top or worth testing?
```
Comment (gbuckingham89) [inline tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php:169]: I'd start by thinking about what the behaviour should be in that situation.

Hopefully, it's not even possible to get into that state.

So in this case, yes, I probably wouldn't have this test.

**Review by gbuckingham89 — CHANGES_REQUESTED** [review]
Overall looks good - just don't need to be sending that magic link as part of the request for the detailed results email.

Once we've got the rest of the overall flow working, we'll come back and do some refactoring I think - we're starting to get a little inconsistent with naming (e.g. the two middleware classes, no sub-directory for the notifications etc).

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php (hunk 3 above)
Comment (jimleeder123) [reply app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:36]: Done and tests changed

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php (hunk 4 above)
Comment (jimleeder123) [reply app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:55]: Have just removed the comment

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php (hunk 5 above)
Comment (jimleeder123) [reply app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:25]: I'm changing it to `$spaceCalculatorInput` to match the model name

File: app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php (hunk 6 above)
Comment (jimleeder123) [reply app/Http/Middleware/GuardSpaceCalculatorDetailedOutput.php:59]: Done thanks

File: app/Http/Controllers/Web/SpaceCalculator/OutputsController.php (hunk 7 above)
Comment (jimleeder123) [reply app/Http/Controllers/Web/SpaceCalculator/OutputsController.php:162]: I think this has come from interpreting the flow diagram incorrectly perhaps. We have the form on the page so the user can choose to authenticate or not manually instead.

File: app/Http/Controllers/Web/SpaceCalculator/OutputsController.php (hunk 8 above)
Comment (jimleeder123) [reply app/Http/Controllers/Web/SpaceCalculator/OutputsController.php:164]: Done

File: tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php (hunk 9 above)
Comment (jimleeder123) [reply tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php:60]: Removed

File: tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php (hunk 10 above)
Comment (jimleeder123) [reply tests/Feature/Http/Web/SpaceCalculator/OutputsController/GetDetailedTest.php:169]: Removed

**Review by gbuckingham89 — APPROVED** [review]
Looks good Jim - nice work 👍

**Pull request merged**